engine/framework/internal/eventloop/runner.go

```diff
@@ -115,6 +122,20 @@ func isForcefulExitError(errIn error) bool {
 		return true
 	}
 
+	if derrors.ErrWorkerCancel.Equal(errIn) {
+		// In canceling scenario business logic promises that all workers have
+		// been canceled, which means we don't need a graceful exit.
+		return true
```

Comment:
I think it' s only for the `jobmaster`.
For `Worker`, things become subtle here. 
DM may hope to maintain the worker's status in `OnWorkerOffline`. If we skip `GracefulExit` here, 
DM job master need to rely on the `OnWorkerStatusUpdate` to update its worker's status.

Reply:
Move the logic to NotifyExit function of job master

Reply:
Ok. But in my mind, `Cancel` is just a normal operation like job finish.
`jobmaster` exiting directly has pros and cons:
* Pros: `jobmaster` can call `StopImpl` quickly
* Cons: `jobmanager` need to wait a timeout to update its jobmaster status